### ISSUE TITLE
[Improvement] Better definition for set_le and set_lt, and possibly multiple ordering definitions

### ISSUE THREAD

**Issue by dimitri-justeau** [MiniZinc/libminizinc#623]
Hello,

With @cprudhom (https://github.com/cprudhom), we are currently questioning the exact ordering definition of `set_le` and `set_lt`. The documentation says: "lexicographic order of the sorted lists of elements". This definition is clear when two sets s1 and s2 have the same cardinality:

{0,1,2} < {1, 2, 3} < {1, 3, 4} < {2, 3, 4}, < {2, 3, 5} etc.

However, when s1 and s2 do not have the same cardinality, I think that it is necessary to define the ordering more precisely. Currently, it seems that MiniZing uses a colexicographic order (https://en.wikipedia.org/wiki/Lexicographic_order#Colexicographic_order) with trailing zeros at the left when s1 and s2 do not have the same cardinality. With this ordering, if we consider all the subsets of {0, 1, 2} we would have:

{} < {0} < {1} < {0, 1} < {2} < {0, 2} < {1, 2} < {0, 1, 2}

However, MiniZinc seems to consider that {0, 1} = {1}, which indicates that zero is treated as a neutral element. This is not necessarily a problem, but I think that it should be clarified, and even better proposed as an option, because I can think of several cases in which we do not want {0, 1} = {1}, e.g. when the set elements correspond to the indices of sorted cost array.

Moreover, to order sets of different cardinalities, several options are possible:

- Use a shortlex ordering, first order by cardinality, then by lexicographic order (https://en.wikipedia.org/wiki/Shortlex_order).
- Use a classical lexicographic ordering (https://en.wikipedia.org/wiki/Lexicographic_order#Motivation_and_definition), with trailing zeros at the right if two sets do not have the same cardinality (e.g. {1, 2, 3} vs {1, 2} -> {1, 2, 3} vs {1, 2, 0}). In that case ordering the subsets of {1, 2, 3} results in : {} < {1} < {1, 2} < {1, 2, 3} < {1, 3} < {2} < {2, 3} < {3}.
- Use a colexicographic order (https://en.wikipedia.org/wiki/Lexicographic_order#Colexicographic_order), which is an inverted lexicographic order, with trailing zeros at the left (e.g. {1, 2, 3} vs {1, 2} -> {1, 2, 3} vs {0, 1, 2}), in that case for {1, 2, 3} we have: {} < {1} < {2} < {1, 2} < {3} < {1, 3} < {2, 3} < {1, 2, 3}.

So in conclusion, I believe that MiniZinc currently uses a lexicographic ordering, with zero treated as a neutral element, and not a lexicographic ordering. Do you believe that it would be possible to offer several options (shortlex, lexicographic, lexicographic)? And also the neutrality of zero optional?

Best regards,

Dimitri.

**Comment by cyderize** [comment]
From what I remember, MiniZinc uses the classical lexicographic ordering (although unfortunately Gecode uses a different ordering, but if we use `nosets.mzn` then `set_lt` will be rewritten using a decomposition with the correct definition).

Running
```
include "nosets.mzn";
array [1..8] of var set of 1..3: x;
constraint forall (i in 1..7) (x[i] < x[i + 1]);
```
Should give the result

```
x = [{}, 1..1, 1..2, 1..3, {1,3}, 2..2, 2..3, 3..3];
```

Can you give a model where MiniZinc is considering that {0, 1} = {1}? That doesn't match what I'd expect to happen.

**Comment by dimitri-justeau** [comment]
Ah ok, my apologies, you are totally right, I misinterpreted the results of the test in https://github.com/chocoteam/choco-solver/issues/936#issuecomment-1282298357 because I expected the ordering to be colexicographically ordered, in which case it would have meant that {0, 1} = {1}. But as you say the order is indeed lexicographic, and thus {0, 1} < {1}. Sorry for this.

We need to fix the constraint in Choco so it matches MiniZinc definition.

Excepted my crazy idea of a neutral zero, I still think that it could be useful to allow for other set ordering definitions. I will suggest the idea to implement it in Choco, at least.

Best regards,

Dimitri.
